src/control/probe.rs

```diff
@@ -0,0 +1,463 @@
+                // SAFETY: This function was made for us, and we uphold our own constraints.
+                return Some(unsafe {
+                    group_probe_items(self.probe.ctrl, self.index, bit, self.probe.bucket_mask)
+                });
+            } else {
+                // SAFETY: The probe always returns an item.
+                let group_probe = unsafe { self.probe.next().unwrap_unchecked() };
+
+                // SAFETY: The hash table contains this type of items.
+                let full = unsafe { group_probe.full::<T>() };
+                self.iter = full.iter;
+                self.index = full.index;
+            }
+        }
+    }
+}
+
+/// Probe which iterates through groups in the control data to search for
+/// buckets for a given hash value.
+///
+/// Sequence based on triangular numbers, which is guaranteed (since our table
+/// size is a power of two) to visit every group of elements exactly once.
+///
+/// A triangular probe has us jump by 1 more group every time. So first we
+/// jump by 1 group (meaning we just continue our linear scan), then 2 groups
+/// (skipping over 1 group), then 3 groups (skipping over 2 groups), and so on.
+///
+/// Proof that the probe will visit every group in the table:
+/// <https://fgiesen.wordpress.com/2015/02/22/triangular-numbers-mod-2n/>
+#[derive(Clone)]
+pub(crate) struct Probe {
+    /// Current position in sequence.
+    pos: usize,
+
+    /// Offset from current position to next group.
+    stride: usize,
+
+    /// Pointer to control data.
+    ctrl: NonNull<Tag>,
+
+    /// Number of buckets, minus one.
+    bucket_mask: usize,
+
+    /// Tag containing the hash of the item.
+    hash_tag: Tag,
+
+    /// Initial hash, for debugging.
+    #[cfg(debug_assertions)]
+    hash: u64,
+
+    /// Current iteration, for debugging.
+    #[cfg(debug_assertions)]
+    iter: usize,
+}
+impl Probe {
+    /// Creates a new probe for a table for a given hash.
+    ///
+    /// # Safety
+    ///
+    /// The given bucket mask must be equal to the power-of-two number of buckets minus one,
+    /// and the given control pointer must be a valid hash table.
+    pub(crate) unsafe fn new(ctrl: NonNull<Tag>, bucket_mask: usize, hash: u64) -> Probe {
+        debug_assert!((bucket_mask + 1).is_power_of_two());
+
+        Probe {
+            // On 32-bit platforms we simply ignore the higher hash bits.
+            //
+            // We need to apply the bucket mask now in the rare case where it
+            // would fall greater than `usize::MAX - Group::WIDTH`, since it
+            // could overflow.
+            pos: (hash as usize) & bucket_mask,
+            hash_tag: Tag::full(hash),
+            stride: 0,
+            ctrl,
+            bucket_mask,
+            #[cfg(debug_assertions)]
+            hash,
+            #[cfg(debug_assertions)]
+            iter: 0,
+        }
+    }
+
+    /// Gets the inner slice of groups.
+    fn control_slice(&self) -> &[Tag] {
```

Comment:
This method basically indicated a bug in the existing group code: we assume that there are always this many initialized tags in the control data, but that's not what `Group::static_empty` gave us: it only gave us a single group, rather than two like this would imply.

So, to help ensure that methods like this always work, and so we don't have to waltz around the concept of "we probably won't read that many groups," the static group allocations were doubled.

Reply:
Actually a single group is sufficient. We only need `GROUP_SIZE - 1` trailing bytes, and the static empty table has 1 bucket. Hence a total of `GROUP_SIZE` control bytes.

Also since `control_slice` is only used by debugging code, it only needs to include the first `bucket_mask + 1` control bytes. The trailing bytes are a mirror of the first bytes, or EMPTY in the case of tables smaller than the group size.

Reply:
So, I agree that we technically only need this many bytes, but all the existing code relies on `num_ctrl_bytes`, which is defined as `bucket_mask + 1 + Group::WIDTH`, not `bucket_mask + Group::WIDTH`. So, in all cases, you need at least two groups.

I'm open to changing this, but I figured that the empty table case wasn't really worth optimising for as long as it didn't allocate anything.

Reply:
So I went back and double-checked the implementation. We only need `Group::WIDTH - 1` tail bytes for probing, but we actually need `Group::WIDTH` padding bytes for `set_ctrl` (the last byte is only written but never read), hence the current definition of `num_ctrl_bytes`.

So the correct length for the static allocation should be `1 + Group::WIDTH`.